### PR TITLE
Make BufferWriter.h, BufferWriterForward.h and MemSpan.h available to C++ plugins.

### PR DESCRIPTION
**Pull request by ywkaras**

### CODE REVIEW

File: lib/ts/Makefile.am [hunk 1]
```diff
@@ -20,7 +20,17 @@ include $(top_srcdir)/build/tidy.mk
...
+# indirectly, which are not intended for direct use by plugins must be included via a symbolic link in the ../ts_no_api
+# directory
+#
+library_include_HEADERS = \
+apidefs.h \
+string_view.h \
+BufferWriter.h \
+BufferWriterForward.h \
+MemSpan.h \
+TextView.h
```
[the remark this answers is not in the document]
Comment (ywkaras) [reply lib/ts/Makefile.am]: Yes, they're copied to include/ts in the install tree.

File: lib/ts/Makefile.am [hunk 2]
```diff
@@ -20,7 +20,17 @@ include $(top_srcdir)/build/tidy.mk
 
 library_includedir=$(includedir)/ts
 
-library_include_HEADERS = apidefs.h string_view.h TextView.h
+# IMPORTANT:  These headers are exported for use by plugings.  Any traffic server headers they in turn include, directly or
```
[the remark this answers is not in the document]
Comment (ywkaras) [reply lib/ts/Makefile.am]: OK, I'll also fix another place this was misspelled that I found with fgrep/find.

File: lib/ts/Makefile.am (hunk 1 above)
[the remark this answers is not in the document]
Comment (d2r) [reply lib/ts/Makefile.am]: OK, white-list sounds good. Nice comment there too.

**Comment by SolidWallOfCode** [comment]
I'm working on cleaning up these headers to remove dependencies on ATS internals. On a first pass, it looks like it is mainly `ink_assert` that is the problem.

I think it's fine to put `ink_std_compat.h` in the normal API area. Plugins may want to use those as well.

**Comment by ywkaras** [comment]
One thing to consider is the issue of any sort of ABI for C++ plugins.  I think, with this change, we would definitely have to require that plugins were built with the same compiler and options as the core.

**Comment by ywkaras** [comment]
In general, the header files for an interface will likely include other header files that should _not_ be part of the interface.  Maybe there's a better way to do it, but I don't think it's something that necessarily can be avoided.

**Comment by zwoop** [comment]
[approve ci clang-format]

**Comment by SolidWallOfCode** [comment]
I think we could do better, if necessary by splitting files so that the exposed interface files can include files that are not inappropriate for plugins.

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
Superseded by #3724.

**Pull request closed without merging**